File: src/mutate4lua/selection.lua
local selection = {}
local function scope_hash_map(scopes)
  local map = {}
  for _, scope in ipairs(scopes or {}) do
    map[scope.id] = scope.semantic_hash
  end
  return map
end
local function mutation_count(sites, scope_ids)
  if not scope_ids then
    return #sites
  end
  local count = 0
  for _, site in ipairs(sites) do
    if scope_ids[site.scope_id] then
      count = count + 1
    end
  end
  return count
end
function selection.changed_scopes(previous_manifest, analysis)
  if not previous_manifest then
    return {
      manifest_exists = false,
      project_hash_changed = false,
      all = {},
      unregistered = {},
      violations = {},
    }
  end
  if previous_manifest.project_hash == analysis.project_hash then
    return {
      manifest_exists = true,
      project_hash_changed = false,
      all = {},
      unregistered = {},
      violations = {},
    }
  end
  local previous = scope_hash_map(previous_manifest.scopes)
  local unregistered = {}
  local violations = {}
  local all = {}
  for _, scope in ipairs(analysis.scopes) do
    local old_hash = previous[scope.id]
    if not old_hash then
      unregistered[scope.id] = true
      all[scope.id] = true
    elseif old_hash ~= scope.semantic_hash then
      violations[scope.id] = true
      all[scope.id] = true
    end
  end
  return {
    manifest_exists = true,
    project_hash_changed = true,
    all = all,
    unregistered = unregistered,
    violations = violations,
  }
end
function selection.filter(args, analysis, previous_manifest, coverage_lines)
  local changed = selection.changed_scopes(previous_manifest, analysis)
  local selected = {}
  local covered = {}
  local uncovered = {}
  local all_sites = analysis.sites
  local function line_selected(site)
    if not args.lines or #args.lines == 0 then
      return true
    end
    return args.lines_lookup[site.line] == true
  end
  local function scope_selected(site)
    if args.mutate_all then
      return true
    end
    if args.lines and #args.lines > 0 then
      return true
    end
    if args.since_last_run then
      return changed.all[site.scope_id] == true
    end
    if not previous_manifest then
      return true
    end
    if not changed.project_hash_changed then
      return false
    end
    return changed.all[site.scope_id] == true
  end
  for _, site in ipairs(all_sites) do
    if line_selected(site) and scope_selected(site) then
      selected[#selected + 1] = site
      if coverage_lines and next(coverage_lines) ~= nil then
        if coverage_lines[site.relative_file .. ":" .. site.line] then
          covered[#covered + 1] = site
        else
          uncovered[#uncovered + 1] = site
        end
      else
        covered[#covered + 1] = site
      end
    end
  end
  return {
    total_mutation_sites = #all_sites,
    selected = selected,
    covered = covered,
    uncovered = uncovered,
    changed_mutation_sites = mutation_count(all_sites, changed.all),
    manifest_exists = changed.manifest_exists,
    project_hash_changed = changed.project_hash_changed,
    differential_surface_area = mutation_count(all_sites, changed.unregistered),
    manifest_violating_surface_area = mutation_count(all_sites, changed.violations),
    changed_scope_ids = changed.all,
  }
end
return selection

File: cmd/mutate4lua-go/selection.go
package main

func scopeHashMap(scopes []scopeInfo) map[string]string {
	result := map[string]string{}
	for _, scope := range scopes {
		result[scope.ID] = scope.SemanticHash
	}
	return result
}

func mutationCount(sites []mutationSite, scopeIDs map[string]bool) int {
	if scopeIDs == nil {
		return len(sites)
	}
	count := 0
	for _, site := range sites {
		if scopeIDs[site.ScopeID] {
			count++
		}
	}
	return count
}

func computeChangedScopes(previous *manifestData, analysis analysisResult) changedScopes {
	if previous == nil {
		return changedScopes{All: map[string]bool{}, Unregistered: map[string]bool{}, Violations: map[string]bool{}}
	}
	if previous.ProjectHash == analysis.ProjectHash {
		return changedScopes{ManifestExists: true, All: map[string]bool{}, Unregistered: map[string]bool{}, Violations: map[string]bool{}}
	}
	previousHashes := scopeHashMap(previous.Scopes)
	all := map[string]bool{}
	unregistered := map[string]bool{}
	violations := map[string]bool{}
	for _, scope := range analysis.Scopes {
		oldHash, ok := previousHashes[scope.ID]
		if !ok {
			unregistered[scope.ID] = true
			all[scope.ID] = true
		} else if oldHash != scope.SemanticHash {
			violations[scope.ID] = true
			all[scope.ID] = true
		}
	}
	return changedScopes{
		ManifestExists:     true,
		ProjectHashChanged: true,
		All:                all,
		Unregistered:       unregistered,
		Violations:         violations,
	}
}

func filterSelection(args selectionArgs, analysis analysisResult, previous *manifestData, coverageLines map[string]bool) selectionResult {
	changed := computeChangedScopes(previous, analysis)
	selected := []mutationSite{}
	covered := []mutationSite{}
	uncovered := []mutationSite{}
	lineSelected := func(site mutationSite) bool {
		if len(args.Lines) == 0 {
			return true
		}
		return args.LinesLookup[site.Line]
	}
	scopeSelected := func(site mutationSite) bool {
		if args.MutateAll {
			return true
		}
		if len(args.Lines) > 0 {
			return true
		}
		if args.SinceLastRun {
			return changed.All[site.ScopeID]
		}
		if previous == nil {
			return true
		}
		if !changed.ProjectHashChanged {
			return false
		}
		return changed.All[site.ScopeID]
	}
	for _, site := range analysis.Sites {
		if !lineSelected(site) || !scopeSelected(site) {
			continue
		}
		selected = append(selected, site)
		if len(coverageLines) > 0 {
			key := normalizeRelativePath(site.RelativeFile) + ":" + itoa(site.Line)
			if coverageLines[key] {
				covered = append(covered, site)
			} else {
				uncovered = append(uncovered, site)
			}
		} else {
			covered = append(covered, site)
		}
	}
	return selectionResult{
		TotalMutationSites:           len(analysis.Sites),
		Selected:                     selected,
		Covered:                      covered,
		Uncovered:                    uncovered,
		ChangedMutationSites:         mutationCount(analysis.Sites, changed.All),
		ManifestExists:               changed.ManifestExists,
		ProjectHashChanged:           changed.ProjectHashChanged,
		DifferentialSurfaceArea:      mutationCount(analysis.Sites, changed.Unregistered),
		ManifestViolatingSurfaceArea: mutationCount(analysis.Sites, changed.Violations),
		ChangedScopeIDs:              changed.All,
	}
}
